### PR TITLE
Add `Self: ~const Trait` to traits with `#[const_trait]`

### PR DESCRIPTION
**Pull request by fee1-dead**
r? @oli-obk

### CODE REVIEW

File: src/test/ui/consts/min_const_fn/cmp_fn_pointers.stderr [hunk 1]
```diff
@@ -1,10 +1,26 @@
...
    |
-   = note: see issue #53020 <https://github.com/rust-lang/rust/issues/53020> for more information
+   = help: the trait `~const PartialEq<_>` is not implemented for `fn()`
+note: the trait `PartialEq<_>` is implemented for `fn()`, but that implementation is not `const`
+  --> $DIR/cmp_fn_pointers.rs:2:16
+   |
+LL |     unsafe { x == y }
+   |                ^^
+   = help: the following other types implement trait `PartialEq<Rhs>`:
+             extern "C" fn() -> Ret
```
Comment (deltragon) [inline src/test/ui/consts/min_const_fn/cmp_fn_pointers.stderr:14]: Do these types implement PartialEq as const? If not, this seems like a slight diagnostic regression. (Not sure if that's relevant for this PR.)

File: src/test/ui/consts/const-eval/const_raw_ptr_ops.stderr [hunk 2]
```diff
@@ -1,18 +1,49 @@
-error: pointers cannot be reliably compared during const eval
```
Comment (oli-obk) [inline src/test/ui/consts/const-eval/const_raw_ptr_ops.stderr:1]: The const checking logic for this message can now just be an assert or a delay_span_bug I believe

**Comment by fee1-dead** [comment]
Note to self: need to prepare a PR for disallowing `impl const Trait`s where `Trait` is not marked with `const_trait`.

**Pull request merged**